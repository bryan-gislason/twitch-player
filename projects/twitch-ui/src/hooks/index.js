export { default as useKeyboard } from "./useKeyboard";
export { default as useLanguage } from "./useLanguage";
export { default as useLayout } from "./useLayout";
export { default as useOverrides } from "./useOverrides";
export { default as useRouter } from "./useRouter";
export { default as useTheme } from "./useTheme";
