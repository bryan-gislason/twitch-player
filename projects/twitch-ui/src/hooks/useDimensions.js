import { throttle } from "@twitch-player/utils";
import { useCallback, useContext } from "react";
import { DimensionsContext } from "../context";

const useDimensions = key => {
  const [dimensions, setDimensions] = useContext(DimensionsContext);

  const updateDimensionsCallback = useCallback(
    throttle(({ nativeEvent }) => {
      setDimensions(prevState => ({ ...prevState, [key]: nativeEvent.layout }));
    }),
    [key]
  );

  return [dimensions[key], updateDimensionsCallback];
};

export default useDimensions;
