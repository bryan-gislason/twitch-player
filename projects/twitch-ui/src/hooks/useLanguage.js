import { useTranslation } from "@twitch-player/i18n";
import { useEffect } from "react";

const useLanguage = onLanguageChanged => {
  const { i18n } = useTranslation();

  const handleLanguageChanged = () => {
    const lng = i18n.language;
    const dir = i18n.dir(lng);

    onLanguageChanged(lng, dir);
  };

  useEffect(() => {
    handleLanguageChanged();

    i18n.on("languageChanged", handleLanguageChanged);

    return () => {
      i18n.off("languageChanged", handleLanguageChanged);
    };
  }, []);
};

export default useLanguage;
