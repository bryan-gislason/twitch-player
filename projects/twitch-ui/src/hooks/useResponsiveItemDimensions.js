import { useCallback, useState } from "react";
import { throttle } from "../utils";

const useResponsiveItemDimensions = ({
  maxItemWidth,
  maxItemHeight,
  spaceSides,
  spaceBetween,
}) => {
  const [layout, updateLayout] = useState(null);
  const width = layout ? layout.width : undefined;

  const numColumns = width ? Math.ceil(width / maxItemWidth) : undefined;
  const itemWidth = width
    ? Math.floor(
        (width - 2 * spaceSides - (numColumns - 1) * 2 * spaceBetween) /
          numColumns
      )
    : undefined;
  const itemHeight = width
    ? Math.floor((itemWidth * maxItemHeight) / maxItemWidth)
    : undefined;

  const updateLayoutCallback = useCallback(
    throttle(({ nativeEvent }) => {
      updateLayout(nativeEvent.layout);
    }),
    []
  );

  return [{ numColumns, itemWidth, itemHeight }, updateLayoutCallback];
};

export default useResponsiveItemDimensions;
