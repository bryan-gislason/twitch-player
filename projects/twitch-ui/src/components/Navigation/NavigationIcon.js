import PropTypes from "prop-types";
import React, { memo } from "react";
import { Image, StyleSheet, View } from "react-native";
import { useTheme } from "../../hooks";
import Hoverable from "../Hoverable";
import InternalTouchableLink from "../InternalTouchableLink";

const NavigationIcon = ({
  href,
  source,
  accessibilityLabel,
  testID = "navigation-icon",
}) => {
  const { colors, layout } = useTheme();

  const imageSize = layout.navSize;
  const imageStyle = {
    height: imageSize,
    margin: layout.gapSmall,
    width: imageSize,
  };

  return (
    <Hoverable>
      {isHover => {
        const wrapper = [
          styles.wrapper,
          {
            backgroundColor: isHover
              ? colors.navBackgroundHover
              : "transparent",
            margin: layout.gapSmall,
          },
        ];

        return (
          <InternalTouchableLink href={href} testID={testID}>
            <View style={wrapper} testID={`${testID}-wrapper`}>
              <Image
                source={{ uri: source }}
                accessibilityLabel={accessibilityLabel}
                style={imageStyle}
                testID={`${testID}-image`}
              />
            </View>
          </InternalTouchableLink>
        );
      }}
    </Hoverable>
  );
};

NavigationIcon.propTypes = {
  href: PropTypes.string.isRequired,
  source: PropTypes.string.isRequired,
  accessibilityLabel: PropTypes.string.isRequired,
  testID: PropTypes.string,
};

const styles = StyleSheet.create({
  wrapper: {
    borderRadius: "50%",
  },
});

export default memo(NavigationIcon);
