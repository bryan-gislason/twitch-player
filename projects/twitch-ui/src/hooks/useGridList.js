import { useTheme } from "@twitch-player/themes/dist/hooks";
import {
  getItemDimensions,
  getGridListItemLayout,
  getGridListItemStyle,
} from "../utils";
import { useDimensions } from ".";

export const useCategoryGrid = itemCount => {
  const [{ layout, typography }] = useTheme();
  const [{ width, height }] = useDimensions("main");

  const options = {
    itemCount,
    containerWidth: width,
    containerHeight: height,
    maxImageWidth: layout.maxCategoryWidth,
    maxImageHeight: layout.maxCategoryHeight,
    detailHeight:
      Math.floor(typography.sizeSecondary * typography.lineHeight) +
      Math.floor(typography.sizeTertiary * typography.lineHeight) +
      layout.gapSmall * 2,
    spaceSides: layout.gapExtraLarge,
    spaceBetween: layout.gapSmall,
    spaceBottom: layout.gapMedium,
  };

  const itemDimensions = getItemDimensions(options);
  const extendedOptions = { ...options, ...itemDimensions };

  const getItemLayout = itemCount && getGridListItemLayout(extendedOptions);
  const renderItemStyle = itemCount && getGridListItemStyle(extendedOptions);

  return {
    ...extendedOptions,
    getItemLayout,
    renderItemStyle,
  };
};

export const useChannelGrid = itemCount => {
  const [{ layout, typography }] = useTheme();
  const [{ width, height }] = useDimensions("main");

  const options = {
    itemCount,
    containerWidth: width,
    containerHeight: height,
    maxImageWidth: layout.maxChannelWidth,
    maxImageHeight: layout.maxChannelHeight,
    detailHeight:
      Math.floor(typography.sizeSecondary * typography.lineHeight) +
      Math.floor(typography.sizeTertiary * typography.lineHeight) +
      Math.floor(typography.sizeTertiary * typography.lineHeight) +
      layout.gapSmall * 2,
    spaceSides: layout.gapExtraLarge,
    spaceBetween: layout.gapSmall,
    spaceBottom: layout.gapMedium,
  };

  const itemDimensions = getItemDimensions(options);
  const extendedOptions = { ...options, ...itemDimensions };

  const getItemLayout = itemCount && getGridListItemLayout(extendedOptions);
  const renderItemStyle = itemCount && getGridListItemStyle(extendedOptions);

  return {
    ...extendedOptions,
    getItemLayout,
    renderItemStyle,
  };
};
