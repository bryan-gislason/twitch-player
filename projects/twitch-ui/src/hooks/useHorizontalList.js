import { useTheme } from "@twitch-player/themes/dist/hooks";
import {
  getItemDimensions,
  getHorizontalListItemLayout,
  getHorizontalListItemStyle,
} from "../utils";
import { useDimensions } from ".";

export const useCategoryList = itemCount => {
  const [{ layout }] = useTheme();
  const [{ width }] = useDimensions("main");

  const options = {
    itemCount,
    containerWidth: width,
    maxImageWidth: layout.maxCategoryWidth,
    maxImageHeight: layout.maxCategoryHeight,
    spaceSides: layout.gapExtraLarge,
    spaceBetween: layout.gapSmall,
    spaceBottom: layout.gapMedium,
  };

  const itemDimensions = getItemDimensions(options);
  const extendedOptions = { ...options, ...itemDimensions };

  const getItemLayout =
    itemCount && getHorizontalListItemLayout(extendedOptions);
  const renderItemStyle =
    itemCount && getHorizontalListItemStyle(extendedOptions);

  return {
    ...extendedOptions,
    getItemLayout,
    renderItemStyle,
  };
};

export const useChannelList = itemCount => {
  const [{ layout }] = useTheme();
  const [{ width }] = useDimensions("main");

  const options = {
    itemCount,
    containerWidth: width,
    maxImageWidth: layout.maxChannelWidth,
    maxImageHeight: layout.maxChannelHeight,
    spaceSides: layout.gapExtraLarge,
    spaceBetween: layout.gapSmall,
    spaceBottom: layout.gapMedium,
  };

  const itemDimensions = getItemDimensions(options);
  const extendedOptions = { ...options, ...itemDimensions };

  const getItemLayout =
    itemCount && getHorizontalListItemLayout(extendedOptions);
  const renderItemStyle =
    itemCount && getHorizontalListItemStyle(extendedOptions);

  return {
    ...extendedOptions,
    getItemLayout,
    renderItemStyle,
  };
};
