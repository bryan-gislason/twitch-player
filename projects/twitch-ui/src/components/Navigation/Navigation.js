import { useTranslation } from "@twitch-player/i18n";
import PropTypes from "prop-types";
import React, { memo } from "react";
import { StyleSheet, View } from "react-native";
import { useLayout, useTheme } from "../../hooks";
import NavigationIcon from "./NavigationIcon";

const homeIcon = require("@twitch-player/assets/dist/ui/icons/home.svg");
const channelsIcon = require("@twitch-player/assets/dist/ui/icons/channels.svg");
const categoriesIcon = require("@twitch-player/assets/dist/ui/icons/categories.svg");
const searchIcon = require("@twitch-player/assets/dist/ui/icons/search.svg");
const settingsIcon = require("@twitch-player/assets/dist/ui/icons/settings.svg");

const layoutQuery = {
  horizontal: {
    minWidth: 301,
  },
  vertical: {
    maxWidth: 300,
  },
};

const Navigation = memo(({ testID = "navigation" }) => {
  const [t] = useTranslation();
  const [layout, updateLayout] = useLayout(layoutQuery);
  const { colors } = useTheme();

  const rootStyle = [
    styles.root,
    styles[layout || "horizontal"],
    {
      backgroundColor: colors.navBackground,
    },
  ];

  return (
    <View
      accessibilityRole="navigation"
      style={rootStyle}
      testID={testID}
      onLayout={updateLayout}
    >
      {layout && (
        <>
          <View style={styles[layout]}>
            <NavigationIcon
              href="/"
              source={homeIcon}
              accessibilityLabel={t("components.navigation.homeLabel")}
              testID={`${testID}-home-icon`}
            />
            <NavigationIcon
              href="/channels"
              source={channelsIcon}
              accessibilityLabel={t("components.navigation.channelsLabel")}
              testID={`${testID}-channels-icon`}
            />
            <NavigationIcon
              href="/categories"
              source={categoriesIcon}
              accessibilityLabel={t("components.navigation.categoriesLabel")}
              testID={`${testID}-categories-icon`}
            />
            <NavigationIcon
              href="/search"
              source={searchIcon}
              accessibilityLabel={t("components.navigation.searchLabel")}
              testID={`${testID}-search-icon`}
            />
          </View>
          <View style={styles[layout]}>
            <NavigationIcon
              href="/settings"
              source={settingsIcon}
              accessibilityLabel={t("components.navigation.settingsLabel")}
              testID={`${testID}-settings-icon`}
            />
          </View>
        </>
      )}
    </View>
  );
});

Navigation.propTypes = {
  testID: PropTypes.string,
};

const styles = StyleSheet.create({
  root: {
    justifyContent: "space-between",
  },
  horizontal: {
    flexDirection: "row",
  },
  vertical: {
    flexDirection: "column",
  },
});

export default Navigation;
