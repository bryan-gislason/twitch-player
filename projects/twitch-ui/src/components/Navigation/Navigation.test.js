import PropTypes from "prop-types";
import React from "react";
import { View } from "react-native";
import { useDimensions } from "../../hooks";
import { fireEvent, render } from "../../testing";
import Navigation from "./Navigation";

jest.mock("./NavigationIcon", () => "mock-navigation-icon");

const FakePage = ({ children }) => {
  const [dimensions, updateDimensions] = useDimensions("page");

  return <View onLayout={updateDimensions}>{dimensions && children}</View>;
};

FakePage.propTypes = {
  children: PropTypes.node.isRequired,
};

describe("Navigation", () => {
  let instance;

  beforeEach(() => {
    instance = render(
      <FakePage>
        <Navigation />
      </FakePage>
    );
  });

  it("renders nothing until layout", () => {
    expect(instance.queryByTestId("navigation")).toBeNull();
  });

  describe("with narrow parent", () => {
    beforeEach(() => {
      fireEvent.resize(window, { target: { width: 480, height: 640 } });
    });

    it("renders with a horizontal layout", () => {
      expect(instance.getByTestId("navigation")).toMatchSnapshot();
    });
  });

  describe("with wide parent", () => {
    beforeEach(() => {
      fireEvent.resize(window, { target: { width: 640, height: 480 } });
    });

    it("renders with a vertical layout", () => {
      expect(instance.getByTestId("navigation")).toMatchSnapshot();
    });
  });
});
