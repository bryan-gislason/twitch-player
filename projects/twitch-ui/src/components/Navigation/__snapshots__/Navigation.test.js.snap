// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation renders nothing until layout 1`] = `
<nav
  class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-18u37iz rn-justifyContent-1wtj0ep rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  data-testid="navigation"
  role="navigation"
  style="background-color: rgb(75, 54, 124);"
/>
`;

exports[`Navigation with narrow parent renders with a horizontal layout 1`] = `
<nav
  class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-justifyContent-1wtj0ep rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  data-testid="navigation"
  role="navigation"
  style="background-color: rgb(75, 54, 124);"
>
  <div
    class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  >
    <mock-navigation-icon
      accessibilitylabel="Home"
      href="/"
      source="some-image.png"
      testid="navigation-home-icon"
    />
    <mock-navigation-icon
      accessibilitylabel="Live Channels"
      href="/channels"
      source="some-image.png"
      testid="navigation-channels-icon"
    />
    <mock-navigation-icon
      accessibilitylabel="Categories"
      href="/categories"
      source="some-image.png"
      testid="navigation-categories-icon"
    />
    <mock-navigation-icon
      accessibilitylabel="Search"
      href="/search"
      source="some-image.png"
      testid="navigation-search-icon"
    />
  </div>
  <div
    class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  >
    <mock-navigation-icon
      accessibilitylabel="Settings"
      href="/settings"
      source="some-image.png"
      testid="navigation-settings-icon"
    />
  </div>
</nav>
`;

exports[`Navigation with wide parent renders with a vertical layout 1`] = `
<nav
  class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-18u37iz rn-justifyContent-1wtj0ep rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  data-testid="navigation"
  role="navigation"
  style="background-color: rgb(75, 54, 124);"
>
  <div
    class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-18u37iz rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  >
    <mock-navigation-icon
      accessibilitylabel="Home"
      href="/"
      source="some-image.png"
      testid="navigation-home-icon"
    />
    <mock-navigation-icon
      accessibilitylabel="Live Channels"
      href="/channels"
      source="some-image.png"
      testid="navigation-channels-icon"
    />
    <mock-navigation-icon
      accessibilitylabel="Categories"
      href="/categories"
      source="some-image.png"
      testid="navigation-categories-icon"
    />
    <mock-navigation-icon
      accessibilitylabel="Search"
      href="/search"
      source="some-image.png"
      testid="navigation-search-icon"
    />
  </div>
  <div
    class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-18u37iz rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  >
    <mock-navigation-icon
      accessibilitylabel="Settings"
      href="/settings"
      source="some-image.png"
      testid="navigation-settings-icon"
    />
  </div>
</nav>
`;
